checker/src/org/checkerframework/checker/initialization/InitializationStore.java

```diff
@@ -154,12 +167,37 @@ protected boolean supersetOf(CFAbstractStore<V, S> o) {
 
     @Override
     public S leastUpperBound(S other) {
+        // Remove invariant annotated fields for optimization
+        Map<FlowExpressions.FieldAccess, V> oldFieldValues = fieldValues;
+        fieldValues = new HashMap<>(fieldValues);
+        for (FieldAccess invariantField : invariantFields.keySet()) {
+            fieldValues.remove(invariantField);
+        }
+        Map<FlowExpressions.FieldAccess, V> oldOtherFieldValues = other.fieldValues;
+        other.fieldValues = new HashMap<>(other.fieldValues);
+        for (FieldAccess invariantField : other.invariantFields.keySet()) {
+            other.fieldValues.remove(invariantField);
+        }
+
         S result = super.leastUpperBound(other);
 
+        // Restore fieldValues
+        fieldValues = oldFieldValues;
+        other.fieldValues = oldOtherFieldValues;
+
         // Set intersection for initializedFields.
         result.initializedFields.addAll(other.initializedFields);
         result.initializedFields.retainAll(initializedFields);
 
+        // Set intersection for invariantFields.
+        for (Entry<FieldAccess, V> e : invariantFields.entrySet()) {
+            if (other.invariantFields.containsKey(e.getKey())) {
+                result.invariantFields.put(e.getKey(), e.getValue());
+            }
+        }
+        // Add invariant annotation again.
+        result.fieldValues.putAll(result.invariantFields);
+
```

Comment:
Can you go through the other methods that use `initializedFields` and either also handle `invariantFields` or add a comment why this is not needed?
E.g. for `superSetOf` is some logic missing or do we not need to do anything, because it is just an optimization?

For `internalVisualize` you could just add `invariantFields` to the output.